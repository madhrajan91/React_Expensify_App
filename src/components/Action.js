import React from 'react'
//Action functional vs class 
const Action = (props) => {
    return (
        <div>
                <button onClick={props.handlePick}
                    disabled={!props.hasOptions}
                >
                    What should I do
                </button>
            </div>
    )
}
/*
class Action extends React.Component {
    constructor(props) {
        super(props);
        this.hasOptions = this.props.hasOptions;
    }
    render() {
        return (
            <div>
                <button onClick={this.props.handlePick}
                disabled={!this.hasOptions}
                >What should I do</button>
            </div>
        );
    }
}
*/
export default Action;