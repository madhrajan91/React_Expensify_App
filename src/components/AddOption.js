import React from 'react';

// add Option
export default class AddOption extends React.Component {
    constructor(props) {
        super(props)
        this.handleSubmitOption = this.handleSubmitOption.bind(this);
        this.state = {
            "error": undefined
        }
    }
    handleSubmitOption (e) {
        e.preventDefault();
        const option = e.target.elements.option.value.trim();

        // passing option upstream since handleAddOption is coming from
        // the <InDecisionApp />
        const error = this.props.handleAddOption(option);
        if (error) {
            // this.setState ( () => {
            //     return {
            //         "error": error
            //     };
            // });

            // or
            this.setState (() => ({"error": error}))
        }
        // if (!error) {
        //     e.target.elements.option.value = "";
        // }
        //}
    }
    render() {
        return (
            <div>
                {this.state.error && <p>ERROR {this.state.error}</p>}
                <form onSubmit={this.handleSubmitOption}>
                    <input type="text" name="option" />
                    <button> Add Option </button>
                </form>
            </div>
        );
    }
}
