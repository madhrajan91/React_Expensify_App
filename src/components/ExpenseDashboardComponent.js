import React from 'react';
import ConnectedExpenseListComponent from './ExpenseListComponent';
import ExpenseListFiltersComponent from './ExpenseListFiltersComponent';
import ExpenseSummaryComponent from './ExpenseSummaryComponent';


const ExpenseDashboardComponent = () => (
    <div>
        This is from my dashboard component
        <ExpenseListFiltersComponent />
        <ConnectedExpenseListComponent />
        <ExpenseSummaryComponent />
    </div>
)

export default ExpenseDashboardComponent;