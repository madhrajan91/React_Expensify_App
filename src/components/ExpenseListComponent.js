import React from 'react'
import {connect} from 'react-redux';

import selectFilteredExpenses from './../selectors/expenses';

import ExpenseListItemComponent from './ExpenseListItemComponent';

// export for testing (you don't want to test the connector)
export const ExpenseList = (props) => (
    <div>
        <h1> Expense List </h1>
        {
            props.expenses.length === 0 ? (<p>No Expenses</p>):
            (props.expenses.map((expense) => (
                <ExpenseListItemComponent key={expense.id} expense={expense} />
            )))
        }


        {props.expenses.length}
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses:selectFilteredExpenses(state.expenses, state.filters)

    }
}
export default  connect(mapStateToProps)(ExpenseList);


// const ConnectedExpenseList = connect((state) => {
//     return {
//         expenses:state.expenses
//     }
// })(ExpenseList);

// export default ConnectedExpenseList;