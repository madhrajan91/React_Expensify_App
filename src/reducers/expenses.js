// Expenses REducer
const expensesReducerDefaultState = []
const expensesReducer = (state= expensesReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_EXPENSE':
            //return state.concat(action.expense); // don't use push since push adds to the original while concat returns a new object (Could   USE SPREAD instead)
            return [... state, action.expense]; // SPREAD
        case 'REMOVE_EXPENSE':
            const expenses = state.filter((expenseToRemove) => {
                return (expenseToRemove.id != action.id);
            })
            return expenses;
        case 'EDIT_EXPENSE':
            return state.map((expense)=> {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                }
                else {
                    return expense;
                }
            })
        default:
            return state;
    }
}

export default expensesReducer;