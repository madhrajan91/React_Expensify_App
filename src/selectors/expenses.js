const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    const filteredExpenses = expenses.filter((expense) => {
       console.log(text)
       const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
       const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
       const textMatch = text === '' || expense.description.toLowerCase().includes(text.toLowerCase()) ;

       return startDateMatch && endDateMatch && textMatch;
   })
       console.log(sortBy)
       return filteredExpenses.sort((a,b) => {
           if (sortBy === 'date') {
               return a.createdAt > b.createdAt?1:-1;
           }
           else {
               return a.amount > b.amount?1:-1;
           }
       })

   //return expenses;
}

export default getVisibleExpenses;