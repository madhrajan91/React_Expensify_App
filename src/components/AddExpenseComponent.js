import React from 'react';
import {connect} from 'react-redux'

import ExpenseFormComponent from './ExpenseFormComponent'
import {addExpense} from '../actions/expenses'

const AddExpenseComponent = (props) => (
    <div>
        Add expense
        <ExpenseFormComponent
         onSubmit = {(expense) => {

            console.log(expense)
            props.dispatch(addExpense(expense));
            props.history.push('/')
         }} />
    </div>
)


export default connect()(AddExpenseComponent);