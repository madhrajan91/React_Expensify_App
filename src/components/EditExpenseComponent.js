import React from 'react';

import {connect} from 'react-redux';

import ExpenseFormComponent from './ExpenseFormComponent';
import {editExpense} from '../actions/expenses'
import {removeExpense} from './../actions/expenses';

const EditExpenseComponent = (props) => {
    console.log(props);
    // query strings 
    // hash
    console.log(props.location.search);
    console.log(props.location.hash);
    return (
        <div>
            Edit  with id of {props.match.params.id}
            <ExpenseFormComponent onSubmit={(expense)=> {
                console.log(expense.id);
    
                props.dispatch(editExpense(props.match.params.id, expense));
                props.history.push('/')
            }}
            expense={props.expense}
            />
            <button onClick= {(e) => {
                props.dispatch(removeExpense(props.expense));
                console.log('removed');
                props.history.push('/')
            }}>Remove</button>
        </div>
    )
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            if (expense.id === props.match.params.id){
                return true;
            }
            else {
                return false;
            }
        })
    }
}

export default connect(mapStateToProps)(EditExpenseComponent);