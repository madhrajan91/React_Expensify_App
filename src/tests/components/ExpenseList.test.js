import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseList} from '../../components/ExpenseListComponent';
import expenses from '../fixtures/expenses';

test ('should render expenselist from expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot()
})

test ('should render expenselist with empty expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={[]}/>);
    expect(wrapper).toMatchSnapshot()
})