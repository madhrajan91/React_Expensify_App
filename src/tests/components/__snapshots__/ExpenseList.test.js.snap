// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenselist from expenses 1`] = `
<div>
  <h1>
     Expense List 
  </h1>
  <ExpenseListItemComponent
    expense={
      Object {
        "amount": 100,
        "createdAt": 1000,
        "description": "Water bill",
        "id": 1,
        "note": "",
      }
    }
    key="1"
  />
  <ExpenseListItemComponent
    expense={
      Object {
        "amount": 20,
        "createdAt": 5000,
        "description": "Gum bill1",
        "id": 2,
        "note": "",
      }
    }
    key="2"
  />
  <ExpenseListItemComponent
    expense={
      Object {
        "amount": 40,
        "createdAt": -345600000,
        "description": "Gum bill2",
        "id": 3,
        "note": "",
      }
    }
    key="3"
  />
  <ExpenseListItemComponent
    expense={
      Object {
        "amount": 500,
        "createdAt": 864000000,
        "description": "insurance ",
        "id": 4,
        "note": "",
      }
    }
    key="4"
  />
  4
</div>
`;

exports[`should render expenselist with empty expenses 1`] = `
<div>
  <h1>
     Expense List 
  </h1>
  <p>
    No Expenses
  </p>
  0
</div>
`;
